README.md
# Router

A SPA routing library for lazy loading Web Component with functionality similar to [FastRoute](https://github.com/nikic/FastRoute#defining-routes) and [ExpressJS](http://expressjs.com/en/guide/routing.html).

## Install

Install via NPM:

```bash
npm i -S @codewithkyle/router
```

Or via CDN:

```javascript
import {
    router,
    navigateTo,
    mount,
    pageJump,
    replaceState,
    pushState,
    enableTransition,
    disableTransition,
    setTransitionTimer,
    transition
} from "https://cdn.jsdelivr.net/npm/@codewithkyle/router@2/router.min.mjs";
```

```html
<script src="https://cdn.jsdelivr.net/npm/@codewithkyle/router@2/router.min.js">
```

## Usage

```typescript
// routes.js

import {
    router,
    navigateTo,
    mount,
    pageJump,
    replaceState,
    pushState,
    transition
} from "https://cdn.jsdelivr.net/npm/@codewithkyle/router@2/router.min.mjs";

// Enable page transitions
router.enableTransitions();

// Disable page transitions (default)
router.disableTransitions();

// Override auto page transition timer (defaults to 5000)
router.setTransitionTimer(600) // ms

// Disable auto page transition timer
router.setTransitionTimer(-1); // accepts -1, null, Infinity

// Add a route to a custom file (supports external URLs)
router.add("/", {
    tagName: "homepage-component",
    file: "./homepage.js",
});

// Automatically load and mount the demo-page web component from the './demo-page.js' file
router.add("/", "demo-page");

// Routes now support closures
router.add("/closure", (tokens, params) => {
    sessionStorage.setItem("closure", "true");
    alert("You may now access the test pages.");
});

// Routes now support redirecs
router.redirect("/dead-link", "/");

// Create router groups with a prefix and optional middleware (middleware can be an array of functions)
router.group(
    {
        prefix: "/blog",
        middleware: (tokens, params) => {
            if (!sessionStorage.getItem("closure")) {
                alert("Access blocked until you run the closure.");

                // Throw a URL to force a redirect
                throw location.origin;
            }
        },
    },
    (router) => {
        router.redirect("/dead-link", "/");

        // Chain groups to extend prefixs or add additional middleware closures
        router.group({ prefix: "/article" }, (router) => {
            // Route tokens now support RegExp strings (anything after the ':' character)
            router.add("/{SLUG:\\d+}", "blog-number");

            // Route tokens without a RegExp string default to /.*/ (anything)
            router.add("/{SLUG}", "blog-article");
        });
    }
);

// Routes are now checked in the order that they're created
// Add a wildcard (*) route at the bottom to catch any route
router.add("/*", "missing-page");
```

```typescript
// homepage.js

// You can export your Web Components as default or as a named export.
export default class Homepage extends HTMLElement {
    constructor(tokens: Tokens, params: Params) {
        super();
        // ...snip...
    }
}
```

### Loading Animation

Since this library bypasses the navive browser navigation functionality you will need to create your own loading state. When loading a route the Router will set a `[router]` attribute on the `<HTML>` element. You can use the snippets blow to create a custom loading animation.

When page transitions (`router.enableTransitions()`) is in use you can trigger page transitions in two ways:

1. You can set a default auto transition timer using `router.setTransitionTimer(ms)`
1. You can manually trigger the transition using the `transition()` method

Ideally you should use both of these methods. When you are ready to transition call the `transition()` method but also set up a reasonable auto transition timer. The page transition will occur when one of the two triggers resolve.

#### CSS

```css
html[router="loading"] * {
    cursor: wait !important;
}
```

#### SCSS

```scss
.my-class {
    color: blue;

    & html[router="loading"] {
        color: grey;
    }
}
```

### Custom Events

```typescript
type Data = {
    [key:string]: any;
};
type RedirectingDetails = {
    path: string,
    hash: string,
    params: Params,
}
type PreloadingDetails = {
    path: string,
    hash: string,
    params: Params,
};
type LoadingDetails = {
    path: string,
    hash: string,
    params: Params,
    tokens: Tokens,
    data: Data,
};
type LoadedDetails = {
    path: string,
    hash: string,
    tokens: Tokens,
    params: Params,
    data: Data,
};
type OutgoingDetails = {
    path: string,
    hash: string,
    params: Params,
    tokens: Tokens,
};

// Fired after the router has started and is ready to hijack navigation events
document.addEventListener("router:ready", () => {
    console.log("Ready");
});

// Fired when the page has started the routing process
document.addEventListener("router:preloading", (e) => {
    const incoming = e.detail.incoming instanceof PreloadingDetails;
    const outgoing = e.detail.outgoing instanceof OutgoingDetails;
    console.log("Preloading", incoming, outgoing);
});

// Fired when the page has started the loading process
document.addEventListener("router:loading", (e) => {
    const incoming = e.detail.incoming instanceof LoadingDetails;
    const outgoing = e.detail.outgoing instanceof OutgoingDetails;
    console.log("Loading", incoming, outgoing);
});

// Fired after the page has loaded and the history state has been updated
document.addEventListener("router:loaded", (e) => {
    const incoming = e.detail.incoming instanceof LoadedDetails;
    const outgoing = e.detail.outgoing instanceof OutgoingDetails;
    console.log("Loaded", incoming, outgoing);
});

// Fired when the page was redirected
document.addEventListener("router:redirecting", (e) => {
    const incoming = e.detail.incoming instanceof RedirectingDetails;
    const outgoing = e.detail.outgoing instanceof OutgoingDetails;
    console.log("Redirecting", incoming, outgoing);
});
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.1.1] - 2023-03-22

### Fixed

- token parsing ([#15](https://github.com/codewithkyle/router/issues/15))
- anchor hijacking ([#15](https://github.com/codewithkyle/router/issues/15))
- infinite failed module import loop ([#16](https://github.com/codewithkyle/router/issues/16))

## [2.1.0] - 2022-07-26

### Added

- asynchronous middleware and closure support ([#13](https://github.com/codewithkyle/router/issues/13))
- router event details ([#12](https://github.com/codewithkyle/router/issues/12))
- `router:preloading` event
- `router:redirecting` event
- animation and page transition methods ([#11](https://github.com/codewithkyle/router/issues/11))
    - `transition()`
    - `router.enableTransitions()`
    - `router.disableTransitions()`
    - `router.setTransitionTimer(ms)` (defaults to 5000)

## [2.0.0] - 2022-03-12

Version 2 of this router has been redesigned to mimic some of the functionality of [FastRoute](https://github.com/nikic/FastRoute#defining-routes) and [ExpressJS](http://expressjs.com/en/guide/routing.html).

```javascript
// Add a route to a custom file (supports external URLs)
router.add("/", {
    tagName: "homepage-component",
    file: "./homepage.js",
});

// Automatically load and mount the demo-page web component from the './demo-page.js' file
router.add("/", "demo-page");

// Routes now support closures
router.add("/closure", (tokens, params) => {
    sessionStorage.setItem("closure", "true");
    alert("You may now access the test pages.");
});

// Routes now support redirecs
router.redirect("/dead-link", "/");

// Create router groups with a prefix and optional middleware (middleware can be an array of functions)
router.group(
    {
        prefix: "/blog",
        middleware: (tokens, params) => {
            if (!sessionStorage.getItem("closure")) {
                alert("Access blocked until you run the closure.");

                // Throw a URL to force a redirect
                throw location.origin;
            }
        },
    },
    (router) => {
        router.redirect("/dead-link", "/");

        // Chain groups to extend prefixs or add additional middleware closures
        router.group({ prefix: "/article" }, (router) => {
            // Route tokens now support RegExp strings (anything after the ':' character)
            router.add("/{SLUG:\\d+}", "blog-number");

            // Route tokens without a RegExp string default to /.*/ (anything)
            router.add("/{SLUG}", "blog-article");
        });
    }
);

// Routes are now checked in the order that they're created
// Add a wildcard (*) route at the bottom to catch any route
router.add("/*", "missing-page");
```

### Added

-   `router` export
-   `router.add(route, module, middleware = null)` method
-   `router.redirect(route, url, middleware = null)` method
-   `router.group(settings, closuer)` method
-   route token RegExp support ([#6](https://github.com/codewithkyle/router/issues/6))
-   routes support closures ([#8](https://github.com/codewithkyle/router/issues/8))
-   routes are now checked in the order that they're created
-   router now supports middleware closures ([#9](https://github.com/codewithkyle/router/issues/9))
-   router now supports prefixing ([#9](https://github.com/codewithkyle/router/issues/9))

### Removed

-   `configure()` method

## [1.1.1] - 2022-02-17

### Fixed

-   external link bug
-   missing TypeScript definitions ([#7](https://github.com/codewithkyle/router/issues/7))

## [1.1.0] - 2021-05-22

### Fixed

-   cleaned up console logs ([#2](https://github.com/codewithkyle/routing/issues/2))
-   fixed homepage popstate navigation hijacking bug ([#1](https://github.com/codewithkyle/routing/issues/1))

### Added

-   custom events fired on the `document` ([#4](https://github.com/codewithkyle/routing/issues/4))

## [1.0.0] - 2021-03-27

### Added

-   `mount()` functionality
-   `configure()` functionality
-   `navigateTo()` functionality
-   `pageJump()` functionality
-   `replaceState()` functionality
-   `pushState()` functionality
-   declaritative dynamic routing
-   automatic lazy loading

[unreleased]: https://github.com/codewithkyle/router/compare/v2.1.1...HEAD
[2.1.1]: https://github.com/codewithkyle/router/compare/v2.1.0...v2.1.1
[2.1.0]: https://github.com/codewithkyle/router/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/codewithkyle/router/compare/v1.1.1...v2.0.0
[1.1.1]: https://github.com/codewithkyle/router/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/codewithkyle/router/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/codewithkyle/router/releases/tag/v1.0.0


test/404.js
export class MissingPage extends HTMLElement{
    constructor(){
        super();
        this.innerHTML = "404 - Page not found";
    }
}


test/jap.js
export class JustAPage extends HTMLElement{
    constructor(){
        super();
        this.innerHTML = "Just a page.";
    }
}


test/homepage.js
export class Homepage extends HTMLElement{
    constructor(){
        super();
        this.innerHTML = "Homepage";
    }
}

router.d.ts
export type Route = {
    middleware: Array<Function>;
    redirect?: string;
    tagName?: string;
    file?: string;
    regex: Array<RegExp>;
    tokens: Array<string>;
    closure?: Function;
    route: string;
    segments: Array<string>;
};

export type RedirectingDetails = {
    path: string,
    hash: string,
    params: Params,
};
export type PreloadingDetails = {
    path: string,
    hash: string,
    params: Params,
};
export type LoadingDetails = {
    path: string,
    hash: string,
    params: Params,
    tokens: Tokens,
    data: Data,
};

export type LoadedDetails = {
    path: string,
    hash: string,
    tokens: Tokens,
    params: Params,
    data: Data,
};

export type OutgoingDetails = {
    path: string,
    hash: string,
    params: Params,
    tokens: Tokens,
};

export type Module = {
    tagName: string;
    file: string;
};

export type Tokens = {
    [token: string]: string;
};

export type Params = {
    [param: string]: string | Array<string>;
};

export type Data = {
    [key:string]: any;
};

export type GroupSettings = {
    prefix?: string;
    middleware?: Array<Function> | Function;
};

declare class RouterGroup {
    public group(
        settings: GroupSettings,
        closure: (router: Router | RouterGroup) => void
    ): RouterGroup;
    public add(
        route: string,
        module: string | Function | Module,
        middleware?: Function | Array<Function>
    ): void;
    public redirect(
        route: string,
        url: string,
        middleware?: Array<Function>
    ): void;
}

declare class Router {
    public group(
        settings: GroupSettings,
        closure: (router: Router | RouterGroup) => void
    ): RouterGroup;
    public add(
        route: string,
        module: string | Function | Module,
        middleware?: Function | Array<Function>
    ): void;
    public redirect(
        route: string,
        url: string,
        middleware?: Array<Function>
    ): void;
    public enableTransitions(): void;
    public disableTransitions(): void;
    public setTransitionTimer(ms:number|null): void;
}

declare const mount: (element: HTMLElement) => void;
declare const navigateTo: (url: string, history?: "replace" | "push") => void;
declare const pageJump: (hash: string, jump?: "auto" | "smooth") => void;
declare const replaceState: (url: string) => void;
declare const pushState: (url: string) => void;
declare const router: Router;
declare const transition: () => void;


src/router-group.ts
export default class RouterGroup {
    private router: Router;
    private prefix: string;
    private middleware: Array<Function>;

    constructor(router: Router, prefix = "", middleware = []) {
        this.router = router;
        this.prefix = prefix;
        this.middleware = middleware;
    }

    public addMiddleware(closure: Function): void {
        this.middleware.push(closure);
    }

    public appendPrefix(prefix: string): void {
        this.prefix += `/${prefix.trim().replace(/^\/|\/$/g, "")}`;
    }

    public group(settings: GroupSettings, closure: Function) {
        const routerGroup = new RouterGroup(this.router, this.prefix, [
            ...this.middleware,
        ]);
        if (settings?.prefix?.length) {
            routerGroup.appendPrefix(settings.prefix);
        }
        if (settings?.middleware) {
            if (Array.isArray(settings.middleware)) {
                for (let i = 0; i < settings.middleware.length; i++) {
                    routerGroup.addMiddleware(settings.middleware[i]);
                }
            } else {
                routerGroup.addMiddleware(settings.middleware);
            }
        }
        closure(routerGroup);
    }

    public add(route: string, module: string | Function | Module): void {
        const cleanRoute = `${this.prefix}/${route
            .trim()
            .replace(/^\/|\/$/g, "")}`;
        this.router.add(cleanRoute, module, [...this.middleware]);
    }

    public redirect(
        route: string,
        url: string,
        middleware: Array<Function> = null
    ): void {
        const cleanRoute = `${this.prefix}/${route
            .trim()
            .replace(/^\/|\/$/g, "")}`;
        this.router.redirect(cleanRoute, url, [...this.middleware]);
    }
}


src/router.ts
import type { Route, Tokens, Params, GroupSettings, Module, Data, LoadedDetails, LoadingDetails, PreloadingDetails, RedirectingDetails } from "../router";
import RouterGroup from "./router-group";

class Router {
    public routes: Array<Route>;
    private mountingPoint: HTMLElement;
    private modules: {
        [tagName: string]: any;
    };
    private lastRoute: {
        path: string,
        hash: string,
        tokens: Tokens,
        params: Params,
    };
    private useTransitions:boolean;
    private autoTransitionTimer: number|null; // ms
    private transitionPromises: {
        resolve: Function,
        timeoutId: number|null,
    };

    constructor() {
        this.routes = [];
        this.mountingPoint = null;
        this.modules = {};
        this.lastRoute = null;
        this.autoTransitionTimer = 5000;
        this.useTransitions = false;
        this.transitionPromises = {
            resolve: () => {},
            timeoutId: null,
        };
        window.addEventListener("click", this.hijackClick, {
            capture: true,
        });
        window.addEventListener("popstate", this.hijackPopstate, {
            capture: true,
        });
    }

    public enableTransitions(): void {
        this.useTransitions = true;
    }

    public disableTransitions():void {
        this.useTransitions = false;
    }

    /**
     * Amount of time (milliseconds) that can pass before automatically resolving the transition.
     * Accepts any number greater or equal to -1, null, or Infinity.
     * Values of -1, null, or Infinity will prevent the automatic transition resolution. This is NOT recommended.
    */
    public setTransitionTimer(ms:number|null): void {
        if (ms < -1){
            throw "Timer value out of bounds. Value must be greater than or equal to -1.";
        }
        this.autoTransitionTimer = ms;
    }

    public transition(){
        if (this.transitionPromises.timeoutId !== null){
            clearTimeout(this.transitionPromises.timeoutId);
        }
        this.transitionPromises.resolve(); 
    }

    public group(settings: GroupSettings, closure: Function) {
        const routerGroup = new RouterGroup(this);
        if (settings?.prefix?.length) {
            routerGroup.appendPrefix(settings.prefix);
        }
        if (settings?.middleware) {
            if (Array.isArray(settings.middleware)) {
                for (let i = 0; i < settings.middleware.length; i++) {
                    routerGroup.addMiddleware(settings.middleware[i]);
                }
            } else {
                routerGroup.addMiddleware(settings.middleware);
            }
        }
        closure(routerGroup);
    }

    public add(
        route: string,
        module: string | Function | Module,
        middleware: Function | Array<Function> = null
    ): void {
        const routeModel = this.prepareRouteModel(route);
        if (typeof module === "string") {
            routeModel.tagName = module;
            if (module.indexOf("http") === 0) {
                routeModel.file = module;
            } else {
                routeModel.file = `./${module
                    .trim()
                    .replace(/(\.js)$/, "")
                    .trim()}.js`;
            }
        } else if (module instanceof Function) {
            routeModel.closure = module;
        } else {
            routeModel.tagName = module.tagName;
            routeModel.file = module.file;
        }
        if (middleware instanceof Function) {
            routeModel.middleware.push(middleware);
        } else if (Array.isArray(middleware)) {
            routeModel.middleware = [...routeModel.middleware, ...middleware];
        }
        this.routes.push(routeModel);
    }

    public redirect(
        route: string,
        url: string,
        middleware: Array<Function> = null
    ): void {
        const routeModel = this.prepareRouteModel(route);
        if (middleware instanceof Function) {
            routeModel.middleware.push(middleware);
        } else if (Array.isArray(middleware)) {
            routeModel.middleware = [...routeModel.middleware, ...middleware];
        }
        routeModel.redirect = url.trim().replace(/^\/|\/$/g, "");
        this.routes.push(routeModel);
    }

    private prepareRouteModel(route: string): Route {
        const routeModel: Route = {
            tokens: [],
            regex: [],
            middleware: [],
            route: route.trim().replace(/^\/|$\//g, ""),
            segments: [],
        };
        const tokens = routeModel.route.match(/\{.*?\}/g) ?? [];
        for (let i = 0; i < tokens.length; i++) {
            if (tokens[i].indexOf(":") !== -1) {
                const token = tokens[i]
                    .replace(/\{|\}/g, "")
                    .trim()
                    .replace(/\:.*/, "")
                    .trim();
                routeModel.tokens.push(token);
                const regexString = tokens[i]
                    .replace(/^\{|\}$/g, "")
                    .trim()
                    .replace(/.*?\:/, "")
                    .trim();
                routeModel.regex.push(new RegExp(regexString));
                routeModel.route = routeModel.route.replace(tokens[i], `:${i}`);
            } else {
                routeModel.tokens.push(tokens[i].replace(/\{|\}/g, "").trim());
                routeModel.regex.push(new RegExp(/.*/));
                routeModel.route = routeModel.route.replace(tokens[i], `:${i}`);
            }
        }
        routeModel.route = routeModel.route
            .trim()
            .replace(/\/+/g, "/")
            .toLowerCase();
        routeModel.segments = routeModel.route.split("/");
        return routeModel;
    }

    public mount(element: HTMLElement): void {
        this.mountingPoint = element;
        this.route(location.href, "replace");
        this.dispatchEvent("ready");
    }

    public navigateTo(url: string, history: "replace" | "push" = "push"): void {
        if (
            url.indexOf(location.origin) === 0 ||
            url.indexOf("/") === 0 ||
            url.indexOf("#") === 0 || 
            url === "" || 
            url.indexOf("http") === -1
        ) {
            this.route(url, history);
        } else {
            console.error(`Invalid URL: ${url}`);
        }
    }

    public pageJump(hash: string, jump: "auto" | "smooth" = "smooth") {
        const el: HTMLElement = document.body.querySelector(
            `#${hash.replace(/^\#/, "")}`
        );
        if (el) {
            el.scrollIntoView({
                behavior: jump,
                block: "center",
                inline: "center",
            });
        }
        this.replaceState(`${location.origin}${location.pathname}${hash}`);
    }

    public replaceState(url: string): void {
        url = url.replace(location.origin, "").replace(/^\//, "").trim();
        window.history.replaceState(
            {
                url: url,
            },
            document.title,
            `${location.origin}/${url}`
        );
    }

    public pushState(url: string): void {
        url = url.replace(location.origin, "").replace(/^\//, "").trim();
        window.history.pushState(
            {
                url: url,
            },
            document.title,
            `${location.origin}/${url}`
        );
    }

    private hijackPopstate = (e: PopStateEvent) => {
        if ("url" in e.state) {
            this.route(e.state.url, "replace");
        }
    };

    private hijackClick: EventListener = (e: Event) => {
        const el = e.target as HTMLElement;
        const target = el.closest("[href]");
        if (
            target instanceof HTMLAnchorElement &&
            target.target !== "_blank" &&
            target.getAttribute("href") !== null
        ) {
            if (
                target.href.indexOf(location.origin) === 0 ||
                target.href.indexOf("/") === 0 ||
                target.href.indexOf("#") === 0 ||
                target.href === "" ||
                target.href.indexOf("http") === -1
            ) {
                e.preventDefault();
                e.stopPropagation();
                const url = target.getAttribute("href");
                let history = target.getAttribute("history");
                if (history === "push" || history === "replace") {
                    this.route(url, history);
                } else {
                    this.route(url);
                }
            }
        }
    };

    private mountElement(
        el: HTMLElement,
        url: string,
        history: "push" | "replace"
    ): void {
        this.mountingPoint?.firstElementChild?.remove();
        this.mountingPoint.appendChild(el);
        if (history === "replace") {
            this.replaceState(`${location.origin}/${url}`);
        } else {
            this.pushState(`${location.origin}/${url}`);
        }
    }

    private async importModule(url: string): Promise<any> {
        let module = null;
        try {
            module = await import(url);
        } catch (e) {
            console.error(e);
        }
        return module;
    }

    private parseTokens(url: string, route: Route): Tokens {
        const tokens: Tokens = {};
        if (route === null || route.segments[0] === "*"){
            return tokens;
        }
        url = url
            .replace(/\?.*|\#.*/, "")
            .trim()
            .replace(/\/+/g, "/");
        const segments = url.split("/");
        for (let i = 0; i < segments.length; i++) {
            if (route.segments?.[i]?.indexOf(":") !== -1) {
                const index = parseInt(route.segments[i].substring(1));
                tokens[route.tokens[index]] = segments[i];
            }
        }
        return tokens;
    }

    private parseGetParams(url: string): Params {
        const params: Params = {};
        const urlParams = url.match(/\?.*/)?.[0] ?? null;
        if (!urlParams) {
            return params;
        }
        const urlSearchParams = new URLSearchParams(urlParams);
        urlSearchParams.forEach((value, key) => {
            if (!params?.[key]) {
                const values = urlSearchParams.getAll(key);
                if (values.length > 1) {
                    params[key] = values;
                } else {
                    params[key] = values[0];
                }
            }
        });
        return params;
    }

    private async import(
        route: Route,
        url: string,
        tokens: Tokens,
        params: Params,
        data: Data,
    ): Promise<HTMLElement> {
        if (this.modules?.[route.tagName]) {
            return new this.modules[route.tagName].default(tokens, params, data);
        }

        let module = await this.importModule(route.file);
        if (module === null) {
            console.error(`Failed to import module: ${route.file}`);
            return null;
        }

        if (!module?.default) {
            const key = Object.keys(module)?.[0] ?? null;
            if (!key) {
                console.error(`Failed to find module export: ${route.file}`);
                return null;
            }
            module = Object.assign(
                {
                    default: module[key],
                },
                module
            );
        }

        this.modules[route.tagName] = module;

        if (!customElements.get(route.tagName)) {
            customElements.define(route.tagName, module.default);
        }

        return new this.modules[route.tagName].default(tokens, params, data);
    }

    private findRouteModel(url: string): Route {
        let model = null;
        const cleanUrl = url
            .replace(/\?.*|\#.*/g, "")
            .trim()
            .replace(/^\/|\/$/g, "")
            .trim()
            .replace(/\/+/g, "/")
            .toLowerCase();
        const segments = cleanUrl.split("/");
        for (let i = 0; i < this.routes.length; i++) {
            const route = this.routes[i];
            if (
                route.segments.includes("*") ||
                route.segments.length === segments.length
            ) {
                let failed = false;
                for (let s = 0; s < route.segments.length; s++) {
                    if (route.segments[s] === segments[s]) {
                        continue;
                    } else if (route.segments[s].indexOf(":") !== -1) {
                        const index = parseInt(route.segments[s].substring(1));
                        if (route.regex[index].test(segments[s])) {
                            continue;
                        } else {
                            failed = true;
                            break;
                        } 
                    } else if (route.segments[s] === "*") {
                        break;
                    } else {
                        failed = true;
                        break;
                    }
                }
                if (!failed) {
                    model = route;
                    break;
                }
            }
        }
        return model;
    }

    private async route(
        url: string,
        history: "replace" | "push" = "push"
    ): Promise<void> {
        // Clean URL
        url = url
            .replace(location.origin, "")
            .replace(/^\/|\/$/g, "")
            .trim();
        // Begin routing
        const data:Data = {};
        let params:Params = this.parseGetParams(url);
        let tokens:Tokens = {};
        let hash = "";
        let path = `/${url.replace(/(\?|\#).*/, "").trim()}`;
        if (url.indexOf("#") !== -1) {
            hash = url.match(/\#.*/)[0];
        }
        if (url.indexOf("#") === 0) {
            this.pageJump(url);
        } else {
            document.documentElement.setAttribute("router", "loading");
            this.dispatchEvent("preloading", {
                path: path,
                hash: hash,
                params: params,
            });
            try {
                // Begin routing logic
                const route = this.findRouteModel(url);
                tokens = this.parseTokens(url, route);

                // Begin optional transition animation logic
                let transitionPromise = null;
                if (this.useTransitions && this.lastRoute !== null){
                    transitionPromise = new Promise(resolve => {
                        let timeoutId:number|null = null;
                        if (this.autoTransitionTimer !== null && this.autoTransitionTimer !== Infinity && this.autoTransitionTimer !== -1){
                            timeoutId = setTimeout(resolve, this.autoTransitionTimer);
                        }
                        this.transitionPromises = {
                            resolve: resolve,
                            timeoutId: timeoutId,
                        };
                    });
                }

                // Notify devs that we are now loading
                // Dispatched after transition logic to prevent race conditions
                this.dispatchEvent("loading", {
                    path: path.length ? path : "/",
                    hash: hash,
                    params: params,
                    tokens: tokens,
                    data: data,
                });

                // Dead route redirect
                if (route === null) {
                    throw "/404";
                }

                // Routing logic
                if (route.middleware.length) {
                    for (const middleware of route.middleware) {
                        await middleware({...tokens}, {...params}, data);
                    }
                }
                if (route?.redirect != undefined) {
                    // Announce redirect
                    let params:Params = this.parseGetParams(route.redirect);
                    let hash = route.redirect.match(/\#.*/)?.[0] ?? "";;
                    let path = `/${route.redirect.replace(/(\?|\#).*/, "").trim()}`;
                    this.dispatchEvent("redirecting", {
                        path: path,
                        params: params,
                        hash: hash,
                    });
                    this.navigateTo(route.redirect, history);
                    return;
                }
                if (route?.closure) {
                    await route.closure({...tokens}, {...params}, data);
                } else {
                    const el = await this.import(route, url, {...tokens}, {...params}, {...data});

                    if (el === null) {
                        return;
                    }
                    
                    // End transition animation logic
                    if (transitionPromise != null){
                        await transitionPromise;
                    }

                    this.mountElement(el, url, history);
                    if (hash.length) {
                        this.pageJump(hash, "auto");
                    } else {
                        window.scrollTo({ top: 0, left: 0, behavior: "auto" });
                    }
                }

                // Routing finished successfully
                document.documentElement.setAttribute("router", "idling");
                this.dispatchEvent("loaded", {
                    path: path.length ? path : "/",
                    hash: hash,
                    tokens: tokens,
                    params: params,
                    data: data,
                });
                this.lastRoute = {
                    path: path.length ? path : "/",
                    hash: hash,
                    tokens: tokens,
                    params: params,
                };
            } catch (url) {
                let params:Params = this.parseGetParams(url);
                let hash = url.match(/\#.*/)?.[0] ?? "";
                let path = `/${url.replace(/(\?|\#).*/, "").trim()}`;
                this.dispatchEvent("redirecting", {
                    path: path,
                    hash: hash,
                    params: params,
                });
                this.navigateTo(url);
            }
        }
    }

    private dispatchEvent(type: "loading" | "loaded" | "ready" | "preloading" | "redirecting", details:LoadingDetails|LoadedDetails|PreloadingDetails|RedirectingDetails = null) {
        let event:CustomEvent;
        switch(type) {
            case "ready":
                event = new CustomEvent(`router:${type}`);
                break;
            default:
                event = new CustomEvent(`router:${type}`, {
                    detail: {
                        outgoing: this.lastRoute,
                        incoming: details,
                    },
                });
                break;
        }
        document.dispatchEvent(event);
    }
}

const router = new Router();
const navigateTo = router.navigateTo.bind(router);
const mount = router.mount.bind(router);
const pageJump = router.pageJump.bind(router);
const replaceState = router.replaceState.bind(router);
const pushState = router.pushState.bind(router);
const transition = router.transition.bind(router);

export { navigateTo, router, mount, pageJump, replaceState, pushState, transition };
